File: solutions/0047-permutations-ii/permutations-ii.cpp
// Given a collection of numbers that might contain duplicates, return all possible unique permutations.
//
// Example:
//
//
// Input: [1,1,2]
// Output:
// [
//   [1,1,2],
//   [1,2,1],
//   [2,1,1]
// ]
//
//


class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int> > ans;
        
        backtrace(nums, ans, 0);
        return ans;
    }
    
    void backtrace(vector<int> nums, vector<vector<int>>& ans, int begin) {
        if(begin == nums.size()) {
            ans.emplace_back(nums);
            return;
        }
        
        for(int i=begin; i<nums.size(); ++i) {
            // if(i>begin && nums[i] == nums[i-1]) continue;  错误！！
            // if(i > begin && nums[i] == nums[begin]) continue; 
            // if(i != begin && ((nums[i] == nums[begin]) || (nums[i] == nums[i-1]))) continue; 错误！！ 
            int j = i-1;  // nums[i]要和前面的一个都不相同！！否则换过去不就一样的嘛
            while(j >= begin && nums[i] != nums[j]) {
                --j;
            }
            
            if(i == begin || j < begin) {
                swap(nums[i], nums[begin]);
                backtrace(nums, ans, begin+1);
                swap(nums[i], nums[begin]);
            }
        }        
    }
};

File: solutions/0047-permutations-ii/permutations-ii.java
// Given a collection of numbers that might contain duplicates, return all possible unique permutations.
//
// Example:
//
//
// Input: [1,1,2]
// Output:
// [
//   [1,1,2],
//   [1,2,1],
//   [2,1,1]
// ]
//
//


class Solution {
    public List<List<Integer>> permuteUnique(int[] nums){
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if(nums==null || nums.length==0) return res;
        boolean[] used = new boolean[nums.length];
        List<Integer> list = new ArrayList<Integer>();
        // 排序为了更好去重
        Arrays.sort(nums);
        dfs(nums, used, list, res);
        return res;
    }

    public void dfs(int[] nums, boolean[] used, List<Integer> list, List<List<Integer>> res){
        if(list.size() == nums.length) {
            res.add(new ArrayList<Integer>(list));
            return;
        }
        
        for(int i=0; i<nums.length; i++) {
            // 使用过跳过
            if(used[i]) continue;
            // 和前一个元素一样，但前一个未使用过，跳过（也就是前一个元素回溯了）
            if(i>0 && nums[i-1] == nums[i] && !used[i-1]) continue;
            
            used[i] = true;
            list.add(nums[i]);
            dfs(nums, used, list, res);
            used[i] = false;
            list.remove(list.size()-1);
        }
    }
}
